slatedb/src/clock.rs

```diff
@@ -0,0 +1,246 @@
+use std::{
+    fmt::Debug,
+    sync::{
+        atomic::{AtomicI64, Ordering},
+        Arc,
+    },
+    time::{Duration, SystemTime, UNIX_EPOCH},
+};
+
+use tracing::info;
+
+use crate::{config::DurabilityLevel, SlateDBError};
+
+/// Defines the generic clock trait that SlateDB extends for user and system operations.
+pub trait Clock: Debug + Send + Sync {
+    /// Returns a timestamp (typically measured in millis since the unix epoch),
+    /// must return monotonically increasing numbers (this is enforced
+    /// at runtime and will panic if the invariant is broken).
+    ///
+    /// Note that this clock does not need to return a number that
+    /// represents the unix timestamp; the only requirement is that
+    /// it represents a sequence that can attribute a logical ordering
+    /// to actions on the database.
+    fn now(&self) -> i64;
+}
+
+/// Defines the clock that SlateDB will use for user facing operations.
+pub trait UserClock: Clock {
+    /// Returns the current time plus the given duration.
+    fn add(&self, duration: Duration) -> i64 {
```

Comment:
Should this take an i64 instead of a duration? The `now()` for user clocks might not return timestamps. It could be LSNs from a database, offsets from a Kafka topic, and so on.

Reply:
I've removed this function. I think I used it at some point, and then the code change in another refactor that made this unnecessary. We can always add it back later if it's useful.